Info message: bailing out due to 'builtins' having no 'pythran' attribute

I've been using pythran in my personal project for a while.

I'm trying to compile a large [file](https://www.dropbox.com/s/3vcopxkk543ydok/GLNetEISimLib.py?dl=0)
using pythran 0.11.0,

`pythran GLNetEISimLib.py -o GLNetEISimLib.so -Ofast -march=native -v`

and it is saying

> INFO: During constant folding, bailing out due to: module 'builtins' has no attribute 'pythran'

Is this message anything of concern? It seems pythran is going into an infinite loop, because this message is being displayed many times.

I had been using pythran 0.9.9.dev from 2021-01-03, and I don't remember seeing this INFO warning. It took a very long time to compile though (about 4h).

I'm using Python 3.8.10 which comes by default in Ubuntu 20.04.


### **UPDATE**

The compilation with pythran 0.11.0 actually crashed, but succeeded to compile with pythran 0.9.9.dev0.

**Here is the output of the pythran 0.11.0 error with -v parameter:**

> INFO:      sys file exists: /usr/local/lib/python3.8/dist-packages/pythran/pythran.cfg

> INFO: platform file exists: /usr/local/lib/python3.8/dist-packages/pythran/pythran-linux.cfg
> WARNING:     user file does not exist: /home/mgirardi/.pythranrc
> INFO: pythranrc section [pythran] is valid and options are correct
> INFO: pythranrc section [typing] is valid and options are correct
> INFO: pythranrc section [compiler] is valid and options are correct
> INFO: During constant folding, bailing out due to: module 'builtins' has no attribute 'pythran'
> INFO: During constant folding, bailing out due to: module 'builtins' has no attribute 'pythran'
> INFO: During constant folding, bailing out due to: module 'builtins' has no attribute 'pythran'
> INFO: During constant folding, bailing out due to: module 'builtins' has no attribute 'pythran'
> INFO: During constant folding, bailing out due to: module 'builtins' has no attribute 'pythran'
> INFO: During constant folding, bailing out due to: module 'builtins' has no attribute 'pythran'
> INFO: During constant folding, bailing out due to: module 'builtins' has no attribute 'pythran'
> INFO: During constant folding, bailing out due to: module 'builtins' has no attribute 'pythran'
> INFO: During constant folding, bailing out due to: module 'builtins' has no attribute 'pythran'
> INFO: During constant folding, bailing out due to: module 'builtins' has no attribute 'pythran'
> INFO: During constant folding, bailing out due to: module 'builtins' has no attribute 'pythran'
> INFO: During constant folding, bailing out due to: module 'builtins' has no attribute 'pythran'
> WARNING: Exporting function 'save_spk_data' that modifies its List argument. Beware that this argument won't be modified at Python call site
> running build_ext
> DEBUG: new_compiler returns <class 'distutils.unixccompiler.UnixCCompiler'>
> building 'GLNetEISimLib' extension
> INFO: C compiler: x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -fPIC
> 
> INFO: compile options: '-DENABLE_PYTHON_MODULE -D__PYTHRAN__=3 -DPYTHRAN_BLAS_BLAS -I/usr/local/lib/python3.8/dist-packages/pythran -I/usr/local/lib/python3.8/dist-packages/numpy/core/include -I/usr/include/python3.8 -c'
> extra options: '-std=c++11 -fno-math-errno -fvisibility=hidden -fno-wrapv -Wno-unused-function -Wno-int-in-bool-context -Wno-unknown-warning-option -Ofast -march=native'
> INFO: x86_64-linux-gnu-gcc: /tmp/tmp9xoywpql.cpp
> /tmp/tmp9xoywpql.cpp: In member function ‘typename __pythran_GLNetEISimLib::RunSimulation_adapt::type<argument_type0, argument_type1, argument_type2, argument_type3, argument_type4, argument_type5, argument_type6, argument_type7, argument_type8, argument_type9, argument_type10, argument_type11, argument_type12, argument_type13, argument_type14, argument_type15, argument_type16, argument_type17, argument_type18, argument_type19, argument_type20, argument_type21, argument_type22, argument_type23, argument_type24, argument_type25, argument_type26, argument_type27, argument_type28, argument_type29, argument_type30, argument_type31, argument_type32>::result_type __pythran_GLNetEISimLib::RunSimulation_adapt::operator()(argument_type0&&, argument_type1&&, argument_type2&&, argument_type3&&, argument_type4&&, argument_type5&&, argument_type6&&, argument_type7&&, argument_type8&&, argument_type9&&, argument_type10&&, argument_type11&&, argument_type12&&, argument_type13&&, argument_type14&&, argument_type15&&, argument_type16&&, argument_type17&&, argument_type18&&, argument_type19&&, argument_type20&&, argument_type21&&, argument_type22&&, argument_type23&&, argument_type24&&, argument_type25&&, argument_type26&&, argument_type27&&, argument_type28&&, argument_type29&&, argument_type30&&, argument_type31&&, argument_type32&&) const’:
> /tmp/tmp9xoywpql.cpp:6185:72: warning: typedef ‘__type584’ locally defined but not used [-Wunused-local-typedefs]
>  6185 |     typedef container<typename std::remove_reference<__type583>::type> __type584;
>       |                                                                        ^~~~~~~~~
> x86_64-linux-gnu-gcc: fatal error: Killed signal terminated program cc1plus
> compilation terminated.
> WARNING: Compilation error, trying hard to find its origin...
> Compilation error, trying hard to find its origin...
> WARNING: Nop, I'm going to flood you with C++ errors!
> Nop, I'm going to flood you with C++ errors!
> CRITICAL: Cover me Jack. Jack? Jaaaaack!!!!
> E: error: Command "x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -fPIC -DENABLE_PYTHON_MODULE -D__PYTHRAN__=3 -DPYTHRAN_BLAS_BLAS -I/usr/local/lib/python3.8/dist-packages/pythran -I/usr/local/lib/python3.8/dist-packages/numpy/core/include -I/usr/include/python3.8 -c /tmp/tmp9xoywpql.cpp -o /tmp/tmpvmlgpicy/tmp/tmp9xoywpql.o -std=c++11 -fno-math-errno -fvisibility=hidden -fno-wrapv -Wno-unused-function -Wno-int-in-bool-context -Wno-unknown-warning-option -Ofast -march=native" failed with exit status 1
> Cover me Jack. Jack? Jaaaaack!!!!
> E: error: Command "x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -fPIC -DENABLE_PYTHON_MODULE -D__PYTHRAN__=3 -DPYTHRAN_BLAS_BLAS -I/usr/local/lib/python3.8/dist-packages/pythran -I/usr/local/lib/python3.8/dist-packages/numpy/core/include -I/usr/include/python3.8 -c /tmp/tmp9xoywpql.cpp -o /tmp/tmpvmlgpicy/tmp/tmp9xoywpql.o -std=c++11 -fno-math-errno -fvisibility=hidden -fno-wrapv -Wno-unused-function -Wno-int-in-bool-context -Wno-unknown-warning-option -Ofast -march=native" failed with exit status 1


And here is the output of the pythran 0.9.9.dev0 with -v:

> INFO:      sys file exists: /usr/local/lib/python3.8/dist-packages/pythran/pythran.cfg
> INFO: platform file exists: /usr/local/lib/python3.8/dist-packages/pythran/pythran-linux.cfg
> WARNING:     user file does not exist: /home/mgirardi/.pythranrc
> INFO: pythranrc section [pythran] is valid and options are correct
> INFO: pythranrc section [typing] is valid and options are correct
> INFO: pythranrc section [compiler] is valid and options are correct
> WARNING: Exporting function 'save_spk_data' that modifies its List argument. Beware that this argument won't be modified at Python call site
> running build_ext
> DEBUG: new_compiler returns <class 'distutils.unixccompiler.UnixCCompiler'>
> building 'GLNetEISimLib' extension
> INFO: C compiler: x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -fPIC
> 
> INFO: compile options: '-DENABLE_PYTHON_MODULE -D__PYTHRAN__=3 -DPYTHRAN_BLAS_BLAS -I/usr/local/lib/python3.8/dist-packages/pythran -I/usr/local/lib/python3.8/dist-packages/numpy/core/include -I/usr/include/python3.8 -c'
> extra options: '-std=c++11 -fno-math-errno -w -fvisibility=hidden -fno-wrapv -Ofast -march=native'
> INFO: x86_64-linux-gnu-gcc: /tmp/tmpires7i2y.cpp
> INFO: x86_64-linux-gnu-g++ -pthread -shared -Wl,-O1 -Wl,-Bsymbolic-functions -Wl,-Bsymbolic-functions -Wl,-z,relro -g -fwrapv -O2 /tmp/tmpqvir8ebt/tmp/tmpires7i2y.o -L/usr/lib -o /tmp/tmp4sk9k4dl/GLNetEISimLib.cpython-38-x86_64-linux-gnu.so -fvisibility=hidden -Wl,-strip-all -Ofast -march=native
> INFO: Generated module: GLNetEISimLib
> INFO: Output: GLNetEISimLib.so

--- a/pythran/analyses/__init__.py
+++ b/pythran/analyses/__init__.py
@@ -41,6 +41,6 @@ from .pure_expressions import PureExpressions
 from .range_values import RangeValues
 from .scope import Scope
 from .static_expressions import StaticExpressions, HasStaticExpression
-from .use_def_chain import DefUseChains, UseDefChains
+from .use_def_chain import DefUseChains, UseDefChains, ExtendedDefUseChains
 from .use_omp import UseOMP
 from .yield_points import YieldPoints

--- a/pythran/analyses/aliases.py
+++ b/pythran/analyses/aliases.py
@@ -8,6 +8,7 @@ from pythran.unparse import Unparser
 from pythran.conversion import demangle
 import pythran.metadata as md
 from pythran.utils import isnum
+from pythran.syntax import PythranSyntaxError
 
 import gast as ast
 from copy import deepcopy

--- a/pythran/analyses/ast_matcher.py
+++ b/pythran/analyses/ast_matcher.py
@@ -17,7 +17,7 @@ class Placeholder(AST):
     """ Class to save information from ast while check for pattern. """
 
     def __init__(self, identifier, type=None):
-        """ Placehorder are identified using an identifier. """
+        """ Placeholder are identified using an identifier. """
         self.id = identifier
         self.type = type
         super(Placeholder, self).__init__()

--- a/pythran/tables.py
+++ b/pythran/tables.py
@@ -2164,12 +2164,7 @@ def partialsum(seq):
 
 
 _operator_add_signature = Union[
-    _numpy_binary_op_signature.__args__ +
-    (Fun[[str, str], str],
-     Fun[[List[T0], List[T0]], List[T0]],) +
-    tuple(Fun[[Tuple[t0], Tuple[t1]], Tuple[t0 + t1]]
-          for t0 in partialsum([T0, T1, T2, T3])
-          for t1 in partialsum([T4, T5, T6, T7]))
+    _numpy_binary_op_signature.__args__
 ]
 
 _operator_eq_signature = Union[
@@ -4362,8 +4357,8 @@ MODULES = {
         "is_not": ConstFunctionIntr(),
         "abs": ConstFunctionIntr(),
         "__abs__": ConstFunctionIntr(),
-        "add": ConstFunctionIntr(signature=_operator_add_signature),
-        "__add__": ConstFunctionIntr(signature=_operator_add_signature),
+        "add": ConstFunctionIntr(signature=_numpy_binary_op_signature),
+        "__add__": ConstFunctionIntr(signature=_numpy_binary_op_signature),
         "and_": ConstFunctionIntr(),
         "__and__": ConstFunctionIntr(),
         "floordiv": ConstFunctionIntr(signature=_numpy_binary_op_signature),

--- a/pythran/tests/test_optimizations.py
+++ b/pythran/tests/test_optimizations.py
@@ -458,6 +458,26 @@ def foo(a):
         self.check_ast(init, ref, ["pythran.optimizations.ConstantFolding",
                                    "pythran.optimizations.PatternTransform"])
 
+    def test_lambda_patterns0(self):
+        init = """
+def foo(a):
+    return lambda x, y: x + y"""
+        ref = """import operator as __pythran_import_operator
+def foo(a):
+    return __pythran_import_operator.add"""
+        self.check_ast(init, ref, ["pythran.transformations.RemoveLambdas"])
+
+    def test_lambda_patterns1(self):
+        init = """
+def foo(a):
+    return (lambda x, y: x + 1), (lambda z, w: z + 1)"""
+        ref = """def foo(a):
+    return (foo_lambda0, foo_lambda0)
+def foo_lambda0(x, y):
+    return (x + 1)"""
+        self.check_ast(init, ref, ["pythran.transformations.RemoveLambdas"])
+
+
     def test_inline_builtins_broadcasting0(self):
         init = """
 import numpy as np

--- a/pythran/transformations/remove_lambdas.py
+++ b/pythran/transformations/remove_lambdas.py
@@ -1,14 +1,68 @@
 """ RemoveLambdas turns lambda into regular functions.  """
 
 from pythran.analyses import GlobalDeclarations, ImportedIds
+from pythran.analyses import Check
+from pythran.analyses import ExtendedDefUseChains
 from pythran.passmanager import Transformation
 from pythran.tables import MODULES
 from pythran.conversion import mangle
 
 import pythran.metadata as metadata
 
-from copy import copy
+from copy import copy, deepcopy
 import gast as ast
+
+unaryops = {
+    ast.Invert: 'invert',
+    ast.Not: 'not_',
+    ast.UAdd: 'pos',
+    ast.USub: 'neg',
+}
+
+binaryops = {
+    ast.Add: "add",
+    ast.Sub: "sub",
+    ast.Mult: "mul",
+    ast.Div: "truediv",
+    ast.Mod: "mod",
+    ast.Pow: "pow",
+    ast.LShift: "lshift",
+    ast.RShift: "rshift",
+    ast.BitOr: "or_",
+    ast.BitXor: "xor",
+    ast.BitAnd: "and_",
+    ast.MatMult: "matmul",
+    ast.FloorDiv: "floordiv",
+}
+
+
+def issimpleoperator(node):
+    if node.args.defaults:
+        return None
+    body = node.body
+    args = node.args.args
+
+    if isinstance(body, ast.UnaryOp) and len(args) == 1:
+        if not isinstance(body.operand, ast.Name):
+            return None
+        return unaryops[type(body.op)]
+    if isinstance(body, ast.BinOp) and len(args) == 2:
+        if not all(isinstance(op, ast.Name) for op in (body.left, body.right)):
+            return None
+        if body.left.id != args[0].id or body.right.id != args[1].id:
+            return None
+        return binaryops[type(body.op)]
+
+
+def issamelambda(pattern, f1):
+    f0, duc = pattern
+    if len(f0.args.args) != len(f1.args.args):
+        return False
+    for arg0, arg1 in zip(f0.args.args, f1.args.args):
+        arg0.id = arg1.id
+        for u in duc.chains[arg0].users():
+            u.node.id = arg1.id
+    return Check(f0, {}).visit(f1)
 
 
 class _LambdaRemover(ast.NodeTransformer):
@@ -17,21 +71,28 @@ class _LambdaRemover(ast.NodeTransformer):
         super(_LambdaRemover, self).__init__()
         self.prefix = prefix
         self.parent = parent
+        self.patterns = parent.patterns
 
     def __getattr__(self, attr):
         return getattr(self.parent, attr)
 
     def visit_Lambda(self, node):
-        if MODULES['functools'] not in self.global_declarations.values():
-            import_ = ast.Import([ast.alias('functools', mangle('functools'))])
-            self.imports.append(import_)
-            functools_module = MODULES['functools']
-            self.global_declarations[mangle('functools')] = functools_module
+        op = issimpleoperator(node)
+        if op is not None:
+            if mangle('operator') not in self.global_declarations:
+                import_ = ast.Import([ast.alias('operator', mangle('operator'))])
+                self.imports.append(import_)
+                operator_module = MODULES['operator']
+                self.global_declarations[mangle('operator')] = operator_module
+            return ast.Attribute(ast.Name(mangle('operator'), ast.Load(),
+                                          None, None),
+                                 op, ast.Load())
 
         self.generic_visit(node)
         forged_name = "{0}_lambda{1}".format(
             self.prefix,
             len(self.lambda_functions))
+
 
         ii = self.gather(ImportedIds, node)
         ii.difference_update(self.lambda_functions)  # remove current lambdas
@@ -41,16 +102,33 @@ class _LambdaRemover(ast.NodeTransformer):
         node.args.args = ([ast.Name(iin, ast.Param(), None, None)
                            for iin in sorted(ii)] +
                           node.args.args)
-        forged_fdef = ast.FunctionDef(
-            forged_name,
-            copy(node.args),
-            [ast.Return(node.body)],
-            [], None, None)
-        metadata.add(forged_fdef, metadata.Local())
-        self.lambda_functions.append(forged_fdef)
-        self.global_declarations[forged_name] = forged_fdef
-        proxy_call = ast.Name(forged_name, ast.Load(), None, None)
+        for patternname, pattern in self.patterns.items():
+            if issamelambda(pattern, node):
+                proxy_call = ast.Name(patternname, ast.Load(), None, None)
+                break
+        else:
+            duc = ExtendedDefUseChains()
+            nodepattern = deepcopy(node)
+            duc.visit(ast.Module([ast.Expr(nodepattern)], []))
+            self.patterns[forged_name] = nodepattern, duc
+
+            forged_fdef = ast.FunctionDef(
+                forged_name,
+                copy(node.args),
+                [ast.Return(node.body)],
+                [], None, None)
+            metadata.add(forged_fdef, metadata.Local())
+            self.lambda_functions.append(forged_fdef)
+            self.global_declarations[forged_name] = forged_fdef
+            proxy_call = ast.Name(forged_name, ast.Load(), None, None)
+
         if binded_args:
+            if MODULES['functools'] not in self.global_declarations.values():
+                import_ = ast.Import([ast.alias('functools', mangle('functools'))])
+                self.imports.append(import_)
+                functools_module = MODULES['functools']
+                self.global_declarations[mangle('functools')] = functools_module
+
             return ast.Call(
                 ast.Attribute(
                     ast.Name(mangle('functools'), ast.Load(), None, None),
@@ -86,6 +164,7 @@ class RemoveLambdas(Transformation):
 
     def visit_Module(self, node):
         self.lambda_functions = list()
+        self.patterns = {}
         self.imports = list()
         self.generic_visit(node)
         node.body = self.imports + node.body + self.lambda_functions

--- a/pythran/types/type_dependencies.py
+++ b/pythran/types/type_dependencies.py
@@ -405,8 +405,8 @@ class TypeDependencies(ModuleAnalysis):
         Return type depend on [foo, bar] or [foo, foobar]
         """
         args = [self.visit(arg) for arg in node.args]
-        func = self.visit(node.func)
-        params = args + [func or []]
+        func = self.visit(node.func) or []
+        params = args + [func]
         return list({frozenset.union(*p) for p in itertools.product(*params)})
 
     @staticmethod

